runtime/src/accounts_index.rs

```diff
@@ -220,9 +221,124 @@ impl<T: 'static + Clone> AccountsIndex<T> {
+        //          `R` -> `A` -> `C` -> `B`
+        // Then `B.ancestors == {R, A, C}`
+        //
+        // Next we call `BankForks::set_root()` at some descendant of `R`, `R_new`,
+        // where `R_new > R`.
+        //
+        // When we squash `R_new`, `max_root` in the AccountsIndex here is now set to `R_new`,
+        // and all nondescendants of `R_new` are pruned.
+        //
+        // Now consider any outstanding references to banks in the system that are descended from
+        // `max_root == R_new`. Take any one of these references and call it `B`. Because `B` is
+        // a descendant of `R_new`, this means `B` was also a descendant of `R`. Thus `B`
+        // must be a member of `R_descendants` because `B` was constructed and added to
+        // BankForks before the `set_root`.
+        //
+        // This means by the guarantees of `R_descendants` described above, because
+        // `R_new` is an ancestor of `B`, and `R < R_new < B`, then B.ancestors.contains(R_new)`.
+        //
+        // Now until the next `set_root`, any new banks constructed from `new_from_parent` will
+        // also have `max_root == R_new` in their ancestor set, so the claim holds for those descendants
+        // as well. Once the next `set_root` happens, we once again update `max_root` and the same
+        // inductive argument can be applied again to show the claim holds.
+
+        // Check that the `max_root` is present in `ancestors`. From the proof above, if
+        // `max_root` is not present in `ancestors`, this means the bank `B` with the
+        // given `ancestors` is not descended from `max_root, which means
+        // either:
+        // 1) `B` is on a different fork or
+        // 2) `B` is an ancestor of `max_root`.
+        // In both cases we can ignore the given ancestors and instead just rely on the roots
+        // present as `max_root` indicates the roots present in the index are more up to date
+        // than the ancestors given.
+        let empty = HashMap::new();
+        let ancestors = if ancestors.contains_key(&max_root) {
+            ancestors
+        } else {
+            /*
+            This takes of edge cases like:
+
+            Diagram 1:
+
+                        slot 0
+                          |
+                        slot 1
+                      /        \
+                 slot 2         |
+                    |       slot 3 (max root)
+            slot 4 (scan)
+
+            By the time the scan on slot 4 is called, slot 2 may already have been
+            cleaned by a clean on slot 3, but slot 4 may not have been cleaned.
+            The state in slot 2 would have been purged and is not saved in any roots.
+            In this case, a scan on slot 4 wouldn't accurately reflect he state when bank 4
+            was frozen. In cases like this, we default to a scan on the latest roots by
+            removing all `ancestors`.
+            */
+            &empty
+        };
```

Comment:
This is the fix, removing it will cause `test_store_scan_consistency_unrooted` to fail